examples/Example13/electrons.cu

```diff
@@ -0,0 +1,341 @@
+// SPDX-FileCopyrightText: 2021 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#include "example13.cuh"
+
+#include <AdePT/BVHNavigator.h>
+#include <fieldPropagatorConstBz.h>
+
+#include <CopCore/PhysicalConstants.h>
+
+#include <G4HepEmElectronManager.hh>
+#include <G4HepEmElectronTrack.hh>
+#include <G4HepEmElectronInteractionBrem.hh>
+#include <G4HepEmElectronInteractionIoni.hh>
+#include <G4HepEmElectronInteractionMSC.hh>
+#include <G4HepEmPositronInteractionAnnihilation.hh>
+// Pull in implementation.
+#include <G4HepEmRunUtils.icc>
+#include <G4HepEmInteractionUtils.icc>
+#include <G4HepEmElectronManager.icc>
+#include <G4HepEmElectronInteractionBrem.icc>
+#include <G4HepEmElectronInteractionIoni.icc>
+#include <G4HepEmElectronInteractionMSC.icc>
+#include <G4HepEmPositronInteractionAnnihilation.icc>
+
+// Compute the physics and geometry step limit, transport the electrons while
+// applying the continuous effects and maybe a discrete process that could
+// generate secondaries.
+template <bool IsElectron>
+static __device__ __forceinline__ void TransportElectrons(Track *electrons, const adept::MParray *active,
+                                                          Secondaries &secondaries, adept::MParray *activeQueue,
+                                                          GlobalScoring *globalScoring,
+                                                          ScoringPerVolume *scoringPerVolume)
+{
+#ifdef VECGEOM_FLOAT_PRECISION
+  const Precision kPush = 10 * vecgeom::kTolerance;
+#else
+  const Precision kPush = 0.;
+#endif
+  constexpr int Charge  = IsElectron ? -1 : 1;
+  constexpr double Mass = copcore::units::kElectronMassC2;
+  fieldPropagatorConstBz fieldPropagatorBz(BzFieldValue);
+
+  int activeSize = active->size();
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const int slot      = (*active)[i];
+    Track &currentTrack = electrons[slot];
+    auto volume         = currentTrack.navState.Top();
+    int volumeID        = volume->id();
+    // the MCC vector is indexed by the logical volume id
+    int lvolID     = volume->GetLogicalVolume()->id();
+    int theMCIndex = MCIndex[lvolID];
+
+    // Init a track with the needed data to call into G4HepEm.
+    G4HepEmElectronTrack elTrack;
+    G4HepEmTrack *theTrack = elTrack.GetTrack();
+    theTrack->SetEKin(currentTrack.energy);
+    theTrack->SetMCIndex(theMCIndex);
+    theTrack->SetOnBoundary(currentTrack.navState.IsOnBoundary());
+    theTrack->SetCharge(Charge);
+    G4HepEmMSCTrackData *mscData = elTrack.GetMSCTrackData();
+    mscData->fIsFirstStep        = currentTrack.initialRange < 0;
+    mscData->fInitialRange       = currentTrack.initialRange;
+
+    // Compute safety, needed for MSC step limit.
+    double safety = 0;
+    if (!currentTrack.navState.IsOnBoundary()) {
+      safety = BVHNavigator::ComputeSafety(currentTrack.pos, currentTrack.navState);
```

Comment:
I see that the safety is cached in the track (good!) but I don't see where the cached value is reused , since it seems to be systematically recomputed here. It seems to me we should do:
```
safety = track->GetSafety();
if (safety < vecgeom::kTolerance)
  safety = BVHNavigator::ComputeSafety(currentTrack.pos, currentTrack.navState);
```

Reply:
No, HepEm needs the accurate safety so it must be recomputed (at least that is what `G4HepEmProcess` does and all MSC models in Geant4). There is no point in caching the value and using more memory, if we cannot use it anyway.

Reply:
So I get from this that MSC is indeed applied only within the safety limit, trying not to get out of the current volume, and limiting the step to basically the safety. I'll wait maybe for @mnovak42 comment on the above discussion. I would claim that even if this is the case, working with the reduced safety would be equivalent from MSC point of view, and if this is true it can get rid of several calls to the actual distance. Anyway, `accurate` safety is not a requirement for geometry, which does the best it can returning an overestimate.

Reply:
MSC is not limiting the step such that it can be only at most safety (we would have difficulties crossing any boundary in that case). However, MSC (at least the currently employed step limit algorithm) limits the step based on some information regarding the safety and contains some optimisation based on the safety as well. 

What we do in G4 is:
- we don't use immediately the safety re-computed in the current pre-step point but instead 
- we try to see if the current step limit, given by all other physics, is inside the updated previous safety 
-  this updated safety (called pre-safety) is the previous pre-step point safety minus the previous geometrical step length and displacement if any)    
- the exact safety is computed only if the current physics step would be longer than the updated safety 
- we can do this in G4 since we can interact with the transportation inside the MSC step limit using the `SafetyHelper`
 
When using G4HepEm, the run-time library do not contain any G4 dependence. Therefore, we need to assume the second case above and recompute the safety at the pre-step point (still in the connection layer to G4) before we would ask G4HepEmRun library to compute the step limit. This do not lead to any differences in the results (e.g. the step limit is eventually exactly the same in G4HepEm and native G4) while causing some overhead (especially that we do not even have the physics step length, given by all other physics, at this point.)

So this is why we do not cache and update the safety by subtracting the step lengths done after the exact computation.

Reply:
>  Anyway, `accurate` safety is not a requirement for geometry, which does the best it can returning an overestimate.

As far as I knew safety should always be an underestimate (<=) to the distance to the closest boundary and never an overestimate. This is why I say above that in theory (according to definition) it could always be zero.